utils/pre-swap-checker/contracts/PreSwapChecker.sol

```diff
@@ -26,69 +30,125 @@ contract PreSwapChecker {
       errorCount++;
     }
 
     // Check expiry
     if (order.expiry < block.timestamp) {
-      errors[errorCount] = 'ORDER_EXPIRED';
+      errors[errorCount] = "ORDER_EXPIRED";
       errorCount++;
     }
 
     if (ISwap(swap).signerNonceStatus(order.signer.wallet, order.nonce) != 0x00) {
-      errors[errorCount] = 'ORDER_TAKEN_OR_CANCELLED';
+      errors[errorCount] = "ORDER_TAKEN_OR_CANCELLED";
       errorCount++;
     }
 
     if (order.nonce < ISwap(swap).signerMinimumNonce(order.signer.wallet)) {
-      errors[errorCount] = 'NONCE_TOO_LOW';
+      errors[errorCount] = "NONCE_TOO_LOW";
       errorCount++;
     }
 
+    // check if ERC721 or ERC20 only amount or id set for sender
+    if (order.sender.kind == ERC20_INTERFACE_ID && order.sender.id != 0) {
+      errors[errorCount] = "SENDER_INVALID_ID";
+      errorCount++;
+    } else if (order.sender.kind == ERC721_INTERFACE_ID && order.sender.amount != 0) {
+      errors[errorCount] = "SENDER_INVALID_AMOUNT";
+      errorCount++;
+    }
+
+    // check if ERC721 or ERC20 only amount or id set for signer
+    if (order.signer.kind == ERC20_INTERFACE_ID && order.signer.id != 0) {
+      errors[errorCount] = "SIGNER_INVALID_ID";
+      errorCount++;
+    } else if (order.signer.kind == ERC721_INTERFACE_ID && order.signer.amount != 0) {
+      errors[errorCount] = "SIGNER_INVALID_AMOUNT";
+      errorCount++;
+    }
+
+
     // Check the order sender
     if (order.sender.wallet != address(0)) {
       // The sender was specified
-      // Check their token balance
-      if (!hasBalance(order.sender)) {
-        errors[errorCount] = 'SENDER_BALANCE';
+      // Check if sender kind interface can correctly check balance
+      if (order.sender.kind == ERC721_INTERFACE_ID && !hasValidERC71Interface(order.sender)) {
+        errors[errorCount] = "SENDER_INVALID_ERC721";
         errorCount++;
+      } else {
+        if (!hasBalance(order.sender)) { // Check their token balance
+          errors[errorCount] = "SENDER_BALANCE";
+          errorCount++;
+        }
+
+        // Check their approval
+        if (!isApproved(order.sender, swap)) {
+          errors[errorCount] = "SENDER_ALLOWANCE";
+          errorCount++;
+        }
       }
 
-      // Check their approval
-      if (!isApproved(order.sender, swap)) {
-        errors[errorCount] = 'SENDER_ALLOWANCE';
+      // Check valid token registry handler
+      if (!hasValidKind(order.sender, swap)) {
+        errors[errorCount] = "SENDER_TOKEN_KIND_UNKNOWN";
         errorCount++;
       }
     }
 
-    // Check the order signer
-    if (!hasBalance(order.signer)) {
-      errors[errorCount] = 'SIGNER_BALANCE';
+    // Check if sender kind interface can correctly check balance
+    if (order.signer.kind == ERC721_INTERFACE_ID && !hasValidERC71Interface(order.signer)) {
+      errors[errorCount] = "SIGNER_INVALID_ERC721";
       errorCount++;
+    } else {
+      if (!hasBalance(order.signer)) {
```

Comment:
We know that, because here we are in the `else` branch, we are looking at a token that is not an invalid ERC721 token - but it could be anything else. hasBalance assumes that the token is a valid ERC20 or ERC721 token - which it might not be. e.g. it could be an ERC1155 or something else. Could we be better to do a large if statement with `hasValidKind` so that checks like hasBalance and hasAllowance arent executed on tokens that then turn out to be invalid kinds anyway?

Reply:
that's fair, i had made some faulty assumptions where I was thinking someone may have accidentally put the wrong kind but everything else would be ok. Having a second thought, have a nested system where we don't do balance or approvals checks until validating the kind seems preferable.